osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs

```diff
@@ -0,0 +1,133 @@
+        /// </summary>
+        protected virtual int SectionLength => 400;
+
+        private double currentSectionPeak = 1; // We also keep track of the peak strain level in the current section.
+
+        private double currentSectionEnd;
+
+        private readonly List<double> strainPeaks = new List<double>();
+
+        /// <summary>
+        /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
+        /// </summary>
+        public sealed override void Process(DifficultyHitObject current)
+        {
+            // The first object doesn't generate a strain, so we begin with an incremented section end
+            if (Previous.Count == 0)
+                currentSectionEnd = Math.Ceiling(current.StartTime / SectionLength) * SectionLength;
+
+            while (current.StartTime > currentSectionEnd)
+            {
+                saveCurrentPeak();
+                startNewSectionFrom(currentSectionEnd);
+                currentSectionEnd += SectionLength;
+            }
+
+            CurrentStrain *= strainDecay(current.DeltaTime);
+            CurrentStrain += StrainValueOf(current) * SkillMultiplier;
+
+            currentSectionPeak = Math.Max(CurrentStrain, currentSectionPeak);
+
+            base.Process(current);
+        }
+
+        /// <summary>
+        /// Saves the current peak strain level to the list of strain peaks, which will be used to calculate an overall difficulty.
+        /// </summary>
+        private void saveCurrentPeak()
+        {
+            strainPeaks.Add(currentSectionPeak);
+        }
+
+        /// <summary>
+        /// Sets the initial strain level for a new section.
+        /// </summary>
+        /// <param name="time">The beginning of the new section in milliseconds.</param>
+        private void startNewSectionFrom(double time)
+        {
+            // The maximum strain of the new section is not zero by default, strain decays as usual regardless of section boundaries.
+            // This means we need to capture the strain level at the beginning of the new section, and use that as the initial peak level.
+            currentSectionPeak = GetPeakStrain(time);
+        }
+
+        /// <summary>
+        /// Retrieves the peak strain at a point in time.
+        /// </summary>
+        /// <param name="time">The time to retrieve the peak strain at.</param>
+        /// <returns>The peak strain.</returns>
+        protected virtual double GetPeakStrain(double time) => CurrentStrain * strainDecay(time - Previous[0].StartTime);
+
+        /// <summary>
+        /// Returns the calculated difficulty value representing all processed <see cref="DifficultyHitObject"/>s.
+        /// </summary>
+        public sealed override double DifficultyValue()
+        {
+            double difficulty = 0;
+            double weight = 1;
+
+            var finalStrainPeaks = strainPeaks.Select(p => p).ToList(); // Copy list so we don't create side effects
```

Comment:
Can skip the `.Select(p => p)`. It doesn't do anything.

Reply:
As mentioned in the code comment, this is to ensure that `DifficultyValue()` doesn't produce side effects.
More: https://github.com/ppy/osu/pull/11407#discussion_r550949827

Reply:
Ah nevermind! I misunderstood what you mean. 😅 
Fixed in 645b0c84d8d1e07b0f9b98c74086ae7144ad9f8e